File: src/main/java/core/procedures/sets/IsSubset.java
package core.procedures.sets;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.Set;

public final class IsSubset extends AFn {

  public IsSubset() {
    super(new FnArgsBuilder().min(2).max(2).mandatory(new Class[] {Set.class, Set.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "subset?";
  }

  @Override
  public Boolean apply2(Object set1, Object set2) {
    return ((Set)set2).containsAll((Set)set1);
  }
}

File: src/main/java/core/procedures/sets/IsSubset.kt
package core.procedures.sets

import core.procedures.AFn
import core.procedures.Arity.Exactly

class IsSubset : AFn<Set<*>?, Boolean>(name = "subset?", isPure = true, arity = Exactly(2),
                                       mandatoryArgsTypes = arrayOf(Set::class.java, Set::class.java)) {

    override operator fun invoke(arg1: Set<*>?, arg2: Set<*>?) = arg2!!.containsAll((arg1!!))
}
